source/language/pulses.rst

```diff
@@ -103,3 +102,166 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
+      // reference `freq` variable from enclosing blocks scope
+      frame d0f = newframe(freq, 0.0);
+
+   }
+
+   defcal x $0 {
+      waveform xp = gaussian(1.0, 160t, 40dt);
+      // References channel, frame, and `new_freq` declared in top-level cal block
+      play(d0, xp, d0f);
+      frame.frequency = new_freq;
+      play(d0, xp, d0f);
+   }
+
+
+Restrictions on defcal bodies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The contents of ``defcal`` bodies are subject to the restriction they must have a definite duration
+known at compile time, regardless of the parameters passed in or the state of the system when
+called. This allows the compiler to properly resolve ``durationof(...)`` calls and
+allows for optimizations. If there is to be control flow in the ``defcal``, each branch of the
+control flow must have definite and equivalent duration resolvable at compile time. Similarly, loops
+must be have a resolvable definite duration at compile time.
+
+For example,  consider the case of a ``reset`` gate. The ``defcal`` for a
+``reset`` gate can be composed of a single if statement, provided each branch
+of the if statement has definite and equivalent duration.
+
+.. code-block:: c
+
+   defcal reset $0 {
+      bit res = // measure qubit $0
+      if (res == 1) {
+         // flip the qubit
+      } else {
+         // delay for an equivalent amount of time
+      }
+   }
+
+Calibrations in practice
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+By their very nature calibrations are transient and unique to a target system.
+They are typically generated by automatic calibration routines that are periodically
+run on the target system, that are in turn bootstrapped from previous calibrations.
+The majority of OpenQASM users will use the default calibrations, however,
+for those that want more control, but do not want to bootstrap calibrations for an entire
+system it is expected that the target system provider will provide an include
+file to the user. This will contain the declaration of the ``defcalgrammar``, constants,
+``defcal``s and other grammar and system specific components such as ``channel``s,
+``waveform``s and ``frame``s in the `OpenPulse defcalgrammar <openpulse.html>`. The user
+may then plugin to the existing calibrations by defining new calibrations, or overwriting
+existing ones by using the same ``channel``s and ``frame``s.
+The example below demonstrates this in practice for a two-qubit,
+cross-resonance device using a ``backend.inc`` include file.
+The name ``backend.inc`` is arbitrary - it's just a file to be included using the
+existing ``include`` mechanism.
+
+.. code-block:: c
+
+   // backend.inc for openpulse two-qubit device
+
+   defcalgrammar "openpulse";
+
+   const q0_freq = 5.0e9;
+   const q1_freq = 5.1e9;
+
+   cal {
+
+      extern drag(complex[size] amp, duration l, duration sigma, float[size] beta) -> waveform;
+      extern gaussian_square(complex[size] amp, duration l, duration square_width, duration sigma) -> waveform;
```

Comment:
The current extern declaration grammar does not allow identifiers so need to remove the identifiers in the declarations:

```
extern drag(complex[size], duration, duration, float[size]) -> waveform;
extern gaussian_square(complex[size], duration, duration, duration) -> waveform;
```

Reply:
I am inclined to change the extern declaration grammar to include identifiers, so that it is consistent with non-extern declarations.